### PR TITLE
[ConstraintSystem] Improve @escaping parameter diagnostics

### PR DESCRIPTION
**Pull request by xedin**
Detect difference in escapiness while matching function types
in the solver and record a fix that suggests to add @escaping
attribute where appropriate.

Also emit a tailored diagnostic when non-escaping parameter
type is used as a type of a generic parameter.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test macOS platform

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -450,6 +453,123 @@ bool NoEscapeFuncToTypeConversionFailure::diagnoseAsError() {
...
+
+  auto choice = getOverloadChoiceIfAvailable(locator);
+  if (!choice)
+    return Type();
+
+  if (auto *fnType = choice->openedType->getAs<FunctionType>()) {
+    const auto &param = fnType->getParams()[paramIdx];
+
+    auto paramType = param.getPlainType();
+    if (auto *typeVar = paramType->getAs<TypeVariableType>()) {
```
Comment (DougGregor) [inline lib/Sema/CSDiagnostics.cpp]: `paramType` could be a `DependentMemberType` with a type variable at its base, which we might also want to diagnose this way. `paramType->isTypeVariableOrMember()` checks for both the `TypeVariableType` and `DependentMemberType`-rooted-at-a-`TypeVariableType` case.

**Review by DougGregor — APPROVED** [review]
This looks fantastic!

File: lib/Sema/CSDiagnostics.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Ah interesting, I'll try to add an example and update this code. Thank you!

File: lib/Sema/CSDiagnostics.cpp [hunk 2]
```diff
@@ -450,6 +453,123 @@ bool NoEscapeFuncToTypeConversionFailure::diagnoseAsError() {
...
+
+          emitDiagnostic(GP->getDecl(),
+                         diag::generic_parameters_always_escaping);
+          return true;
+        }
+      }
+
+      // If there are no generic parameters involved, this could
+      // only mean that parameter is expecting @escaping function type.
+      diagnostic = diag::passing_noescape_to_escaping;
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: It could be an Any perhaps

Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: I actually have a check at the top to make sure that this code only handles the case when convert-to type is a function.

File: lib/Sema/CSDiagnostics.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Done!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please clean test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Pull request merged**